gap/SkeletalFinSetsForCAP.gi

```diff
@@ -898,12 +898,18 @@ end );
 ##
 AddCartesianCoevaluationMorphismWithGivenRange( SkeletalFinSets,
   function ( cat, M, N, HN_MxN )
-    local m, n;
+    local m, n, mn;
     
     m := Length( M );
     n := Length( N );
     
-    return MapOfFinSets( cat, M, List( [ 0 .. m - 1 ], i -> Sum( [ 0 .. n - 1 ], j -> ( i + m * j ) * (m*n)^j ) ), HN_MxN );
+    mn := m * n;
+    
+    if mn = 1 then
+        return IdentityMorphism( cat, M );
+    fi;
+    
+    return MapOfFinSets( cat, M, List( [ 0 .. m - 1 ], i -> ( ( ( i - m - 1 + mn ) * mn - i ) * mn^n - ( i - m ) * mn + i ) / ( mn - 1 )^2 ), HN_MxN );
```

Comment:
A suggestion: You could outsource this to a function `GeometricSum`, which would make the intent more clear. Then you could decide if you want the compiler to resolve this function or not (keeping it as a primitive hides the case distinction and might allow for more high-level optimizations in the future, but of course also prevents low-level optimizations).

If you don't want to do this, please add a comment saying that this is the geometric sum of ..., so one can understand this code more easily.